lib/common/xml.c

```diff
@@ -1367,10 +1343,236 @@ mark_child_moved(xmlNode *old_child, xmlNode *new_parent, xmlNode *new_child,
+ *
+ * For other node types, there is no match.
+ *
+ * \param[in] a  New XML child node (<tt>xmlNode *</tt>)
+ * \param[in] b  Old XML child node and whether to require IDs to match
+ *               (<tt>struct new_child_matches_data *</tt>)
+ *
+ * \retval 0 if \p a matches according to \p b
+ * \retval 1 otherwise
+ *
+ * \note This function is suitable for use with \c g_list_find_custom().
+ */
+static int
+new_child_matches(gconstpointer a, gconstpointer b)
+{
+    const xmlNode *new_child = a;
+    struct new_child_matches_data *data = (struct new_child_matches_data *) b;
+    const xmlNode *old_child = data->old_child;
+    const bool match_ids = data->match_ids;
+
+    if (old_child->type != new_child->type) {
+        return 1;
+    }
+
+    switch (old_child->type) {
+        case XML_COMMENT_NODE:
+            return new_comment_matches(old_child, new_child)? 0 : 1;
+        case XML_ELEMENT_NODE:
+            return new_element_matches(old_child, new_child, match_ids)? 0 : 1;
+        default:
+            return 1;
+    }
+}
+
+/*!
+ * \internal
+ * \brief Find matching XML node pairs between old and new XML's children
+ *
+ * A node that is part of a matching pair is removed from its respective list
+ * (\p *old_children or \p *new_children) and added to \p *matches with its
+ * match.
+ *
+ * \param[in,out]  old_children  Children of old XML
+ * \param[in,out]  new_children  Children of new XML
+ * \param[in,out]  matches       Where to prepend pairs of matching nodes from
+ *                               \p old_children and \p new_children
+ * \param[in]      comments_ids  If \c true, match comments and require element
+ *                               IDs to match; otherwise, skip comments and
+ *                               match elements by name only
+ */
+static void
+find_matching_children(GList **old_children, GList **new_children,
+                       GList **matches, bool comments_ids)
+{
+    // Save next pointer because we unlink iter on match
+    for (GList *iter = *old_children, *next = g_list_next(*old_children);
+         iter != NULL;
+         iter = next, next = g_list_next(next)) {
+
+        xmlNode *old_child = iter->data;
+
+        struct new_child_matches_data data = {
+            .old_child = old_child,
+            .match_ids = comments_ids,
+        };
+        GList *match = NULL;
+
+        if (!comments_ids && (old_child->type == XML_COMMENT_NODE)) {
+            continue;
+        }
+
+        match = g_list_find_custom(*new_children, &data, new_child_matches);
+        if (match == NULL) {
+            continue;
+        }
+
+        /* Steal iter and match from *old_children and *new_children,
+         * respectively, and prepend to *matches.
+         *
+         * Order should eventually be (iter, match), with pairs appearing in
+         * document order. We will reverse the order of *matches in the caller
+         * after finding all matches.
```

Comment:
Sigh. This gives document order within each call, but all "ID changed" matches (from the second call) will come after all "ID same" matches (from the first call).

Reply:
Pushed a draft change but there's a kink to work out

Reply:
Okay, I think I'm finally satisfied.